### DISCUSSION TITLE
How to start with STM32duino - STM32G030F6

### DISCUSSION THREAD

**Discussion by novakkry** [stm32duino/Arduino_Core_STM32#1629] [Q&A] [answered]
Hi! A newbie questions here. I wanna use STM32G030F6 [1] for my project. I don't want to use any development board, but bare  standalone chip with internal oscillator. Now my questions. How do I do it using STM32duino?

I've found out that STM32duino supports STM32G030F6 as it was added last month to the supported boards (generic STM32G030F6 [2]). How do I start now? In Arduino IDE I can find the board and I can compile a code for the board but I have no idea what are, for instance, GPIO pins called in Arduino world. Also, the chip has two hardware serials. If I simply use "Serial.begin(xxxx)" what serial on what pins will it use? How to use the second serial? Does Arduino supports that?
I want to use the internal oscillator so I have fewer components on the board. Do I have to somehow tell to compiler to compile it for the internal oscillator on the specific frequency or I just compile it as is?

I wanna program the STM32G030F6 using SWD with ST-link V2.

I just don't know how to start with STM32duino as I don't fully understand the concept. I've always done projects with Arduino Uno or standalone atmega328p-au.

I'll greatly appriciate any suggestions and advices. Thank you.

Krystof

[1] https://www.st.com/en/microcontrollers-microprocessors/stm32g030f6.html#overview
[2] https://github.com/stm32duino/Arduino_Core_STM32/tree/main/variants/STM32G0xx/G030F6P

**Comment by fpistm** [comment]
Duplicate of https://www.stm32duino.com/viewtopic.php?f=7&t=1442
@ABOSTM could you answer to @novakkry, please?

**Comment by ABOSTM** [comment, marked as answer]
Hi @novakkry,
First there is a wiki for Stm32duino, which is a good starting point with many information:
https://github.com/stm32duino/wiki/wiki

and especially
* https://github.com/stm32duino/wiki/wiki#customization

* https://github.com/stm32duino/wiki/wiki/Add-a-new-variant-%28board%29
   which explains many concepts, and in particular SystemClock_Config().

For bare standalone chip with internal oscillator, using Generic STM32G030F6 is the good choice.

Concerning your specific question around Internal oscillator,
as you can see in generic variant for STM32G030F6
https://github.com/stm32duino/Arduino_Core_STM32/blob/73719e6e0b421463ba527449aa9852792dfb25fd/variants/STM32G0xx/G030F6P/generic_clock.c#L37
The PLL is based on HSI clock source. Knowing that HSI is Internal oscillator (as opposed to HSE which would be external oscillator).

In the same directory:
https://github.com/stm32duino/Arduino_Core_STM32/tree/main/variants/STM32G0xx/G030F6P
you can found different files: generic_clock.c, PeripheralPins.c, PinNamesVar.h, variant_generic.cpp, variant_generic.h
with lots of informations about pins GPIO, ...
Note that you can customize those file for your custom board, or even redefine any "WEAK" symbol.

Concerning serial:
Arduino predefines a default "Serial" object which in your case in USART2
https://github.com/stm32duino/Arduino_Core_STM32/blob/73719e6e0b421463ba527449aa9852792dfb25fd/variants/STM32G0xx/G030F6P/variant_generic.h#L113
using default pins PIN_SERIAL_RX and PIN_SERIAL_TX

But you can use any USART instance by declaring a new object :
https://github.com/stm32duino/wiki/wiki/API#hardwareserial
Have a look at PeripheralPins.c to see which pins are eligible for USART TX/RX

To select SWD with ST-link:
![image](https://user-images.githubusercontent.com/50730894/150171905-79d03cee-9d62-4105-9ceb-21235828f7f5.png)
